webClient/src/app/core/menu-bar/menu-bar.component.ts

```diff
@@ -567,19 +567,51 @@ export class MenuBarComponent implements OnInit, OnDestroy {
     this.editorControl.toggleTree.next();
   }
 
-  closeAll() {
-    let closeAllRef;
-    if (this.fileCount == 0) { //TODO: Enhance such that closeAll not visible if no tabs are open
+  promptToSave(file: ProjectContext): Promise<any>{
+    return new Promise((resolve, reject) => {
+      if(file.changed) {
+        const title = 'Do you want to save the changes you made to \'' + file.name + '\?';
+        const warningMessage = 'Your changes will be lost if you don\'t save them.';
+        let response = this.monacoService.confirmAction(title, warningMessage).subscribe(response => {
+          if(response == true) {
+            // when user selects to save the file and close it
+            let sub = this.monacoService.saveFile(file, file.model.path || this.editorControl.activeDirectory).subscribe(() => {
+            resolve(true);
+            });
+          } else if (response != false) {
+            // when user selects to cancel then do not close any file
+            return; 
```

Comment:
If we're awaiting promptToSave, shouldn't we reject when we don't want to close? Or if not rejecting is your way of not letting closeAll complete and thus it "waits forever" which is kind of like cancelling the close, isn't that bad? Maybe I'm misunderstanding